File: backend-springboot/src/main/java/com/cosmetics/model/Product.java
package com.cosmetics.model;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.Id;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.mongodb.core.index.Indexed;
import org.springframework.data.mongodb.core.mapping.Document;
import org.springframework.data.mongodb.core.mapping.Field;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Map;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
@Document(collection = "products")
public class Product {

    @Id
    private String id;

    private String name;

    @Indexed(unique = true)
    private String slug;

    @Indexed(sparse = true)
    private String sku;

    private String description;

    @Field("short_description")
    private String shortDescription;

    @Field("category_id")
    private String categoryId;

    @Field("brand_id")
    private String brandId;

    @Field("base_price")
    private Double basePrice;

    @Field("sale_price")
    @Builder.Default
    private Double salePrice = 0.0;

    private List<String> images;

    private List<Variant> variants;

    private List<Attribute> attributes;

    private List<String> tags;

    @Field("avg_rating")
    @Builder.Default
    private Double avgRating = 0.0;

    @Field("review_count")
    @Builder.Default
    private Integer reviewCount = 0;

    @Field("sold_count")
    @Builder.Default
    private Integer soldCount = 0;

    @Field("is_active")
    @Builder.Default
    private Boolean isActive = true;

    @Field("is_featured")
    @Builder.Default
    private Boolean isFeatured = false;

    @Field("is_new")
    @Builder.Default
    private Boolean isNew = false;

    @Field("is_best_seller")
    @Builder.Default
    private Boolean isBestSeller = false;

    @Field("in_stock")
    @Builder.Default
    private Boolean inStock = true;

    @CreatedDate
    @Field("created_at")
    private LocalDateTime createdAt;

    @LastModifiedDate
    @Field("updated_at")
    private LocalDateTime updatedAt;

    // --- Embedded Classes ---
    
    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    public static class Variant {
        private String sku;
        private String name;
        private Map<String, String> attributes; // vd: {"size": "50ml", "color": "red"}
        private Double price;
        @Field("sale_price")
        private Double salePrice;
        private List<String> images;
    }

    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    public static class Attribute {
        private String key;   // vd: "Thành phần", "Cách dùng"
        private String value; // vd: "Water, Glycerin...", "Thoa đều lên mặt..."
    }
}

File: backend-nodejs/src/models/Product.js
import mongoose from "mongoose";

// Định nghĩa Sub-schema cho biến thể sản phẩm (vd: Dung tích 50ml, 100ml)
const variantSchema = new mongoose.Schema({
  sku: { type: String, required: true },
  name: { type: String, required: true },
  attributes: { type: Map, of: String },
  price: { type: Number, required: true },
  sale_price: { type: Number, default: 0 },
  stock: { type: Number, default: 0 },
  images: [{ type: String }],
});

const productSchema = new mongoose.Schema(
  {
    name: { type: String, required: true },
    slug: { type: String, required: true, unique: true },
    sku: { type: String, sparse: true },
    description: { type: String },
    short_description: { type: String },
    category_id: { type: mongoose.Schema.Types.ObjectId, ref: "Category" },
    brand_id: { type: mongoose.Schema.Types.ObjectId, ref: "Brand" },
    base_price: { type: Number, required: true },
    sale_price: { type: Number, default: 0 },
    images: [{ type: String }],
    tags: [{ type: String }],
    variants: [variantSchema],
    attributes: [
      {
        key: { type: String },
        value: { type: String },
      },
    ],
    rating: { type: Number, default: 0 },
    review_count: { type: Number, default: 0 },
    sold_count: { type: Number, default: 0 },
    is_active: { type: Boolean, default: true },
    is_featured: { type: Boolean, default: false },
    is_new_product: { type: Boolean, default: false },
    is_best_seller: { type: Boolean, default: false },
    stock: { type: Number, default: 0 },
    in_stock: { type: Boolean, default: true },
  },
  { timestamps: { createdAt: "created_at", updatedAt: "updated_at" } },
);

// Map dữ liệu đầu ra để khớp với ProductCard.jsx và AdminProducts.jsx
productSchema.set("toJSON", {
  transform: (doc, ret) => {
    ret.id = ret._id;
    ret.shortDescription = ret.short_description;
    ret.categoryId = ret.category_id;
    ret.brandId = ret.brand_id;
    ret.basePrice = ret.base_price;
    ret.salePrice = ret.sale_price;
    ret.reviewCount = ret.review_count;
    ret.soldCount = ret.sold_count;
    ret.isActive = ret.is_active;
    ret.isFeatured = ret.is_featured;
    ret.isNew = ret.is_new_product;
    ret.isBestSeller = ret.is_best_seller;
    ret.inStock = ret.in_stock;

    // Map sale_price bên trong từng variant
    if (ret.variants) {
      ret.variants = ret.variants.map((v) => {
        v.salePrice = v.sale_price;
        delete v.sale_price;
        delete v._id; // Ẩn _id của sub-document cho gọn
        return v;
      });
    }

    // Dọn dẹp các key thừa
    delete ret._id;
    delete ret.__v;
    delete ret.short_description;
    delete ret.category_id;
    delete ret.brand_id;
    delete ret.base_price;
    delete ret.sale_price;
    delete ret.review_count;
    delete ret.sold_count;
    delete ret.is_active;
    delete ret.is_featured;
    delete ret.is_new_product;
    delete ret.is_best_seller;
    delete ret.in_stock;

    return ret;
  },
});

export default mongoose.model("Product", productSchema);
